File: sdks/typescript/examples/test-environment.ts
/**
 * Environment & Sandbox Configuration Test
 *
 * Verifies the sandbox environment is set up correctly for normal development:
 *   - Running as root (can install packages, modify system files)
 *   - HOME=/root (standard root home)
 *   - /workspace is a separate persistent disk
 *   - npm, pip, git, python, node all work
 *   - npm install works in /workspace
 *
 * Usage:
 *   npx tsx examples/test-environment.ts
 */

import { Sandbox } from "../src/index";

function green(msg: string) { console.log(`\x1b[32m✓ ${msg}\x1b[0m`); }
function red(msg: string) { console.log(`\x1b[31m✗ ${msg}\x1b[0m`); }
function bold(msg: string) { console.log(`\x1b[1m${msg}\x1b[0m`); }
function dim(msg: string) { console.log(`\x1b[2m  ${msg}\x1b[0m`); }

let passed = 0;
let failed = 0;

function check(desc: string, condition: boolean, detail?: string) {
  if (condition) {
    green(desc);
    passed++;
  } else {
    red(`${desc}${detail ? ` (${detail})` : ""}`);
    failed++;
  }
}

async function main() {
  bold("\n╔══════════════════════════════════════════════════╗");
  bold("║       Environment & Sandbox Config Test          ║");
  bold("╚══════════════════════════════════════════════════╝\n");

  let sandbox: Sandbox | null = null;

  try {
    sandbox = await Sandbox.create({ template: "base", timeout: 120 });
    green(`Created sandbox: ${sandbox.sandboxId}`);
    console.log();

    // ── Test 1: Running as root ─────────────────────────────────
    bold("━━━ Test 1: Running as root ━━━\n");

    const whoami = await sandbox.commands.run("whoami");
    check("Running as root", whoami.stdout.trim() === "root",
      `got "${whoami.stdout.trim()}"`);

    const idResult = await sandbox.commands.run("id -u");
    check("UID is 0", idResult.stdout.trim() === "0",
      `got "${idResult.stdout.trim()}"`);

    // ── Test 2: HOME is /root ───────────────────────────────────
    bold("━━━ Test 2: HOME environment ━━━\n");

    const homeResult = await sandbox.commands.run("echo $HOME");
    check("HOME is /root", homeResult.stdout.trim() === "/root",
      `got "${homeResult.stdout.trim()}"`);

    // ── Test 3: /workspace is a separate disk ───────────────────
    bold("━━━ Test 3: /workspace filesystem ━━━\n");

    const dfResult = await sandbox.commands.run("df -h /workspace | tail -1");
    const dfLine = dfResult.stdout.trim();
    dim(`df output: ${dfLine}`);
    check("/workspace is NOT on /dev/root", !dfLine.startsWith("/dev/root"));
    check("/workspace is mounted", dfLine.includes("/workspace"));

    const wsExists = await sandbox.commands.run("test -d /workspace && echo yes || echo no");
    check("/workspace directory exists", wsExists.stdout.trim() === "yes");
    console.log();

    // ── Test 4: Can install system packages ─────────────────────
    bold("━━━ Test 4: Package installation (apt) ━━━\n");

    const aptResult = await sandbox.commands.run(
      "apt-get update -qq > /dev/null 2>&1 && apt-get install -y -qq cowsay > /dev/null 2>&1 && /usr/games/cowsay moo | head -1",
      { timeout: 30 }
    );
    check("apt install works", aptResult.exitCode === 0,
      `exit code ${aptResult.exitCode}`);
    console.log();

    // ── Test 5: npm install in /workspace ────────────────────────
    bold("━━━ Test 5: npm install in /workspace ━━━\n");

    await sandbox.commands.run("mkdir -p /workspace/npm-test");
    await sandbox.files.write("/workspace/npm-test/package.json", JSON.stringify({
      name: "test",
      version: "1.0.0",
      dependencies: { "is-odd": "3.0.1" }
    }));

    const installResult = await sandbox.commands.run(
      "cd /workspace/npm-test && npm install 2>&1 | tail -1",
      { timeout: 30 }
    );
    dim(`npm install exit code: ${installResult.exitCode}`);
    check("npm install succeeds", installResult.exitCode === 0);

    const nodeModules = await sandbox.commands.run(
      "test -d /workspace/npm-test/node_modules && echo yes || echo no"
    );
    check("node_modules created in /workspace", nodeModules.stdout.trim() === "yes");

    const requireTest = await sandbox.commands.run(
      "cd /workspace/npm-test && node -e \"require('is-odd'); console.log('ok')\"",
    );
    check("Installed package works", requireTest.stdout.trim() === "ok");
    console.log();

    // ── Test 6: pip install ─────────────────────────────────────
    bold("━━━ Test 6: pip install ━━━\n");

    const pipResult = await sandbox.commands.run(
      "pip3 install -q requests 2>/dev/null && python3 -c \"import requests; print(requests.__version__)\"",
      { timeout: 30 }
    );
    check("pip install + import works", pipResult.exitCode === 0 && pipResult.stdout.trim().length > 0,
      `exit code ${pipResult.exitCode}`);
    dim(`requests version: ${pipResult.stdout.trim()}`);
    console.log();

    // ── Test 7: Tools available ─────────────────────────────────
    bold("━━━ Test 7: Development tools ━━━\n");

    const tools = ["git", "python3", "node", "npm", "curl", "jq"];
    for (const tool of tools) {
      const r = await sandbox.commands.run(`which ${tool}`);
      check(`${tool} available`, r.exitCode === 0, r.stdout.trim());
    }
    console.log();

    // Clean up
    await sandbox.commands.run("rm -rf /workspace/npm-test");

  } catch (err: any) {
    red(`Fatal error: ${err.message}`);
    if (err.stack) dim(err.stack);
    failed++;
  } finally {
    if (sandbox) {
      await sandbox.kill();
      green("Sandbox killed");
    }
  }

  bold("========================================");
  bold(` Results: ${passed} passed, ${failed} failed`);
  bold("========================================\n");
  if (failed > 0) process.exit(1);
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: sdks/python/examples/test_environment.py
#!/usr/bin/env python3
"""
Environment & HOME Directory Test

Verifies that HOME=/workspace inside sandboxes so that tools like npm, pip,
git etc. use the NVMe-backed workspace drive for caches and config.

Tests:
  1. HOME is set to /workspace
  2. Tilde (~) expands to /workspace
  3. npm cache dir is under /workspace
  4. npm install writes cache to /workspace (not /root)
  5. Dotfiles and configs use /workspace
  6. /workspace is NVMe-backed

Usage:
  python examples/test_environment.py
"""

import asyncio
import sys

from opensandbox import Sandbox

GREEN = "\033[32m"
RED = "\033[31m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"

passed = 0
failed = 0


def green(msg: str) -> None:
    print(f"{GREEN}✓ {msg}{RESET}")


def red(msg: str) -> None:
    print(f"{RED}✗ {msg}{RESET}")


def bold(msg: str) -> None:
    print(f"{BOLD}{msg}{RESET}")


def dim(msg: str) -> None:
    print(f"{DIM}  {msg}{RESET}")


def check(desc: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        green(desc)
        passed += 1
    else:
        red(f"{desc} ({detail})" if detail else desc)
        failed += 1


async def main() -> None:
    global passed, failed

    bold("\n╔══════════════════════════════════════════════════╗")
    bold("║       Environment & HOME Directory Test          ║")
    bold("╚══════════════════════════════════════════════════╝\n")

    sandbox = None

    try:
        sandbox = await Sandbox.create(template="node", timeout=120)
        green(f"Created sandbox: {sandbox.sandbox_id}")
        print()

        # ── Test 1: HOME is /workspace ──
        bold("━━━ Test 1: HOME environment variable ━━━\n")

        result = await sandbox.commands.run("echo $HOME")
        check("HOME is /workspace", result.stdout.strip() == "/workspace",
              f'got "{result.stdout.strip()}"')

        # ── Test 2: Tilde expansion ──
        bold("━━━ Test 2: Tilde (~) expansion ━━━\n")

        result = await sandbox.commands.run("echo ~")
        check("~ expands to /workspace", result.stdout.strip() == "/workspace",
              f'got "{result.stdout.strip()}"')

        result = await sandbox.commands.run("echo ~/test")
        check("~/test expands to /workspace/test",
              result.stdout.strip() == "/workspace/test",
              f'got "{result.stdout.strip()}"')
        print()

        # ── Test 3: npm cache directory ──
        bold("━━━ Test 3: npm cache directory ━━━\n")

        result = await sandbox.commands.run("npm config get cache")
        npm_cache = result.stdout.strip()
        check("npm cache is under /workspace", npm_cache.startswith("/workspace"),
              f'got "{npm_cache}"')
        check("npm cache is NOT under /root", not npm_cache.startswith("/root"),
              f'got "{npm_cache}"')
        dim(f"npm cache dir: {npm_cache}")
        print()

        # ── Test 4: npm install writes to workspace ──
        bold("━━━ Test 4: npm install uses workspace for cache ━━━\n")

        await sandbox.commands.run("mkdir -p /workspace/npm-test")
        await sandbox.files.write("/workspace/npm-test/package.json",
                                  '{"name":"test","version":"1.0.0","dependencies":{"is-odd":"3.0.1"}}')

        install_result = await sandbox.commands.run(
            "cd /workspace/npm-test && npm install --prefer-offline 2>&1",
            timeout=30)
        dim(f"npm install exit code: {install_result.exit_code}")

        root_check = await sandbox.commands.run(
            "du -sh /root/.npm 2>/dev/null || echo 'no /root/.npm'")
        check("No npm cache in /root/.npm",
              "no /root/.npm" in root_check.stdout,
              f'got "{root_check.stdout.strip()}"')

        workspace_cache = await sandbox.commands.run(
            "test -d /workspace/.npm && echo 'exists' || echo 'missing'")
        check("npm cache exists at /workspace/.npm",
              workspace_cache.stdout.strip() == "exists",
              f'got "{workspace_cache.stdout.strip()}"')
        print()

        # ── Test 5: Dotfiles go to workspace ──
        bold("━━━ Test 5: Dotfiles and configs use /workspace ━━━\n")

        result = await sandbox.commands.run("bash -c 'echo $HOME/.bashrc'")
        check(".bashrc path is /workspace/.bashrc",
              result.stdout.strip() == "/workspace/.bashrc",
              f'got "{result.stdout.strip()}"')

        await sandbox.commands.run("git config --global user.name 'Test User'")
        result = await sandbox.commands.run(
            "git config --global --list --show-origin 2>/dev/null | head -1")
        check("git config is under /workspace",
              "/workspace" in result.stdout,
              f'got "{result.stdout.strip()}"')
        print()

        # ── Test 6: /workspace is on the right filesystem ──
        bold("━━━ Test 6: /workspace is NVMe-backed ━━━\n")

        result = await sandbox.commands.run("df -h /workspace | tail -1")
        df_line = result.stdout.strip()
        dim(f"df output: {df_line}")
        check("/workspace is NOT on /dev/root", not df_line.startswith("/dev/root"),
              f'got "{df_line}"')
        print()

        # Clean up
        await sandbox.commands.run("rm -rf /workspace/npm-test")

    except Exception as e:
        red(f"Fatal error: {e}")
        failed += 1
    finally:
        if sandbox:
            await sandbox.kill()
            green("Sandbox killed")

    # --- Summary ---
    bold("========================================")
    bold(f" Results: {passed} passed, {failed} failed")
    bold("========================================\n")
    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
